variants/frontend-react/app/frontend/tests/HelloWorld.spec.jsx

```diff
@@ -0,0 +1,24 @@
+import { render, screen } from '@testing-library/react';
+import userEvent from '@testing-library/user-event';
+import React from 'react';
+import '@testing-library/jest-dom/extend-expect';
+import HelloWorld from '../components/HelloWorld';
+
+describe('HelloWorld', () => {
+  it('renders personalized greeting', async () => {
```

Comment:
Currently this test is actually testing two things that could be in their own separate tests - I think it would be good to split it up so that it's more focused, as usually that makes for better testing.

I'd generally structure the tests like so:

* `describe 'HelloWorld'`
  * `it 'renders the initial greeting'` (tests that `initialGreeting` is rendered)
  * `describe 'when the greeting is changed'`
    * `it 'renders the new greeting'` (tests that the greeting changes)
  * `describe 'when the greeting is cleared'`
    * `it 'falls back to the initial greeting'` (tests that blanking the input has it fall back to the `initialGreeting`)

I've chucked in that last one as a possible third test that'd be easy to implement into the component to give a good reason for a third test - feel free not to implement it; some good alternatives could be either having the `input` value default to `initialGreeting` (so showcasing how to test that an input field has a default value based off a prop), or adding validation to `input` so that it's not allowed to be blank (meaning you'd showcase how you test a more complex thing like validation errors).

Reply:
> I think `findByText` is actually being used as an assertion, which is fine

In saying that, I think it would be good to try and stick to the `expect` matchers as much as possible, which seems to be possible:

```
describe('Stats', () => {
  describe('when the api request is in progress', () => {
    it('shows a loading state', () => {
      fetchPullRequestStatsMock.mockReturnValue(new Promise(() => undefined));

      const { container } = render(<Stats />);

      expect(container).toHaveTextContent(/Loading/iu);
    });
  });

  describe('when the api request resolves', () => {
    it('shows the stats', async () => {
      // useAsync currently requires the Promise to actually be an instanceof the Promise class
      fetchPullRequestStatsMock.mockImplementation(async () =>
        Promise.resolve({
          numberOfPullRequestsClosed: 3,
          averageTimePullRequestsSpendOpen: 1234
        })
      );

      const { container, queryByText } = render(<Stats />);

      await waitForElementToBeRemoved(() => queryByText(/Loading/iu));

      expect(container).not.toHaveTextContent(/Loading/iu);

      expect(container).toHaveTextContent(/PRs Closed: 3/iu);
      expect(container).toHaveTextContent(/Average Time Spent Open: 1234/iu);
    });
  });
})
```

This'd mean we wouldn't have to loosen our linting rules as much, and let us use `expect.hasAssertions()` for every test.

What do you think?